File: src/main/java/edu/practice/datastructures/linkedlist/CircularLinkedList.java
package edu.practice.datastructures.linkedlist;

import java.util.NoSuchElementException;
import java.util.Objects;

public class CircularLinkedList<E> {

    private static class Node<E> {
        E data;
        Node<E> next;

        Node(E data, Node<E> next) {
            this.data = data;
            this.next = next;
        }
    }

    private int size;
    private Node<E> first;
    private Node<E> last;

    public CircularLinkedList() {// Initialize empty Linked List
    }

    public int size() { return size; }

    public boolean isEmpty() {
        return (size == 0);
    }

    public void addFirst(E item) {
        first = new Node<>(item, first);

        if (first.next == null) { // If the added node is the only node, its next should point to itself
            first.next = first;
        }

        if (last == null) {
            last = first;
        }

        size++;
    }

    public void addLast(E item) {
        final Node<E> newNode = new Node<>(item, null);

        if (last == null) {
            newNode.next = newNode;
            last = newNode;
            first = newNode;
        } else {
            newNode.next = first;
            last.next = newNode;
            last = last.next;
        }

        size++;
    }

    public E getFirst() {
        if (first == null) {
            throw new NoSuchElementException();
        }

        return first.data;
    }

    public E getLast() {
        if (last == null) {
            throw new NoSuchElementException();
        }

        return last.data;
    }

    public E removeFirst() {
        final E item = first.data;

        if (first == first.next) { // 1 element in List
            first = null;
            last = null;
        } else {
            first = first.next;
            last.next = first;
        }

        size--;

        return item;
    }

    public E removeLast() {
        final E item = last.data;

        if (first != last) { // More than one element in the list
            Node<E> currentNode = first;
            while (currentNode.next != last) {
                currentNode = currentNode.next;
            }

            currentNode.next = first;
            last = currentNode;
        } else {
            first = null;
            last = null;
        }

        size--;

        return item;
    }

    public boolean contains(E item) {
        if (first == null) {
            return false;
        }

        Node<E> currentNode = first;

        do {
            if (Objects.equals(item, currentNode.data)) {
                return true;
            }

            currentNode = currentNode.next;
        } while (currentNode != first);
        
        return false;
    }
}

File: src/main/kotlin/datastructures/linkedlist/CircularLinkedList.kt
package edu.practice.datastructures.linkedlist


class CircularLinkedList<E>: LinkedList<E> {

    private data class Node<E>(val data: E, var next: Node<E>? = null)

    private var size = 0

    private var first: Node<E>? = null
    private var last: Node<E>? = null

    override fun size(): Int = size

    override fun isEmpty(): Boolean = (size == 0)

    override fun getFirst(): E {
        return first?.data ?: throw NoSuchElementException("Linked List is Empty")
    }

    override fun getLast(): E {
        return last?.data ?: throw NoSuchElementException("Linked List is Empty")
    }

    override fun addFirst(item: E) {
        val newNode = Node(item, first)
        first = newNode

        if (first?.next == null) { // Initially empty before node added
            first?.next = first
            last = first
        }
        
        size++
    }

    override fun addLast(item: E) {
        val newNode = Node(item)

        if (last == null) { // If list is empty
            first = newNode
        } else {
            last?.next = newNode
        }
        last = newNode

        newNode.next = first
        size++
    }

    override fun removeFirst(): E {
        val item = getFirst()

        if (first == first?.next) {
            last = null
            first = null
        } else {
            first = first?.next
            last?.next = first
        }

        size--
        return item
    }

    override fun removeLast(): E {
        val item = getLast()

        if (first == last) { // Only 1 element in the list
            first = null
            last = null
        } else {
            var currentNode = first
            while (currentNode?.next != last) {
                currentNode = currentNode?.next
            }

            currentNode?.next = first
            last = currentNode
        }

        size--
        return item
    }

    override fun contains(item: E): Boolean {
        if (first == null) {
            return false
        }

        var currentNode = first

        do {
            if (item == currentNode?.data) {
                return true
            }

            currentNode = currentNode?.next
        } while (currentNode != first)

        return false
    }
}